idp_node/mapapp/retriever.py

```diff
@@ -0,0 +1,174 @@
+'''
+Build the strings for myproxy-server to use in myproxy-logon,
+See http://grid.ncsa.illinois.edu/myproxy/man/myproxy-server.config.5.html
+'''
+import argparse
+import ConfigParser
+import logging
+import os
+import sys
+
+import OpenSSL
+import psycopg2
+
+def db_connection(config_parser, esgf_home):
+    pass_file = os.path.join(esgf_home, "config", ".esg_pg_pass")
+    logging.info("Retrieving pg pass from %s", pass_file)
+    with open(pass_file, "rb") as filep:
+        password = filep.read().strip()
+
+    section = "installer.properties"
+    db = config_parser.get(section, "db.database")
+    host = config_parser.get(section, "db.host")
+    port = config_parser.get(section, "db.port")
+    user = config_parser.get(section, "db.user")
+
+    connection_params = (
+        "dbname=%s user=%s password=%s host=%s port=%s" %
+        (db, user, password, host, port)
+    )
+    connection_params_no_pass = (
+        "dbname=%s user=%s host=%s port=%s" %
+        (db, user, host, port)
+    )
+    logging.info(connection_params_no_pass)
+
+    return psycopg2.connect(connection_params)
+
+def build_grouprole(args, config_parser, esgf_home):
+    ''' Build the group and role string for a specified username '''
+    logging.info("Fetching groups and roles for %s", args.username)
+
+    section = "installer.properties"
+    esgf_host = config_parser.get(section, "esgf.host")
+    gateway = "https://{}/esgf-idp/openid/%".format(esgf_host)
+
+    conn = db_connection(config_parser, esgf_home)
+    cur = conn.cursor()
+
+    query = " SELECT g.name, r.name"
+    query += " FROM esgf_security.group AS g, esgf_security.role AS r,"
+    query += " esgf_security.permission AS p, esgf_security.user AS u"
+    query += " WHERE p.user_id=u.id"
+    query += " AND u.username=%(username)s"
+    query += " AND u.openid LIKE %(gateway)s"
+    query += " AND p.group_id=g.id"
+    query += " AND p.role_id=r.id"
+    values = {
+        "username": args.username,
+        "gateway": gateway
+    }
+    logging.info(query, values)
+    cur.execute(query, values)
+    result = cur.fetchall()
+    #         //         group     | role
+    #         //    ----------------+------
+    #         //    CMIP5 Research | User
+    #         //    NASA OBS       | User
+    #         //    ORNL OBS       | User
+    if not result:
+        return None
+    grouprole_elements = []
+    for group, role in result:
+        grouprole_elements.append("group_" + group + "_role_" + role)
+    return ";".join(grouprole_elements)
+
+def get_openid(args, config_parser, esgf_home):
+    ''' Retrieve the OpenID for a specified username '''
+    logging.info("Fetching OpenID for %s", args.username)
+
+    section = "installer.properties"
+    esgf_host = config_parser.get(section, "esgf.host")
+    gateway = "https://{}/esgf-idp/openid/%".format(esgf_host)
```

Comment:
Same question as above

Reply:
It is the "wildcard" character for the LIKE clause in the SQL statements.